src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/ObjectWriter/CodeView/CodeViewTypesBuilder.cs

```diff
@@ -0,0 +1,537 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Buffers;
+using System.Buffers.Binary;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Text;
+
+using ILCompiler.DependencyAnalysis;
+using Internal.JitInterface;
+using Internal.TypeSystem;
+using Internal.TypeSystem.TypesDebugInfo;
+
+using static ILCompiler.ObjectWriter.CodeViewNative;
+using static ILCompiler.ObjectWriter.CodeViewNative.CodeViewPointer;
+using static ILCompiler.ObjectWriter.CodeViewNative.CodeViewPropertyFlags;
+using static ILCompiler.ObjectWriter.CodeViewNative.CodeViewType;
+using static ILCompiler.ObjectWriter.CodeViewNative.LeafRecordType;
+
+namespace ILCompiler.ObjectWriter
+{
+    /// <summary>Builder for the CodeView .debug$T section.<summary>
+    /// <remarks>
+    /// The .debug$T section in CodeView contains type (enum, struct, class)
+    /// descriptions. The section is composed of records that are prefixed
+    /// with their type and length. Each record is assigned a type index
+    /// representing its position in the stream. The type indexes start at
+    /// 0x1000 and increase by one for each record. Elementary types, such
+    /// as uint, short, or float, have preassigned indexes below 0x1000,
+    /// represented by the <see cref="CodeViewType" /> enumeration.
+    ///
+    /// The maximum record size is limited to sizeof(ushort) due to the
+    /// layout of the record header. List records (eg. field list) can
+    /// be split into multiple records that are chained together.
+    /// </remarks>
+    internal sealed class CodeViewTypesBuilder : ITypesDebugInfoWriter
+    {
+        private NameMangler _nameMangler;
+        private SectionWriter _sectionWriter;
+        private int _targetPointerSize;
+
+        private uint _classVTableTypeIndex;
+        private uint _vfuncTabTypeIndex;
+        private List<(string, uint)> _userDefinedTypes = new();
+
+        private uint _nextTypeIndex = 0x1000;
+
+        public IList<(string, uint)> UserDefinedTypes => _userDefinedTypes;
+
+        public CodeViewTypesBuilder(NameMangler nameMangler, TargetArchitecture targetArchitecture, SectionWriter sectionWriter)
+        {
+            _nameMangler = nameMangler;
+            _sectionWriter = sectionWriter;
+            _targetPointerSize = targetArchitecture switch
```

Comment:
`TypeSystemContext.Target.PointerSize` is the same value. Is it hard to use it here?